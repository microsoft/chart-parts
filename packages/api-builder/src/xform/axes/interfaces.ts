/*!
 * Copyright (c) Microsoft. All rights reserved.
 * Licensed under the MIT license. See LICENSE file in the project.
 */

import { SceneFrame } from '../context/SceneFrame'
import { SGMark, Axis, Scale, TickValue } from '@chart-parts/interfaces'

export interface AxisSpace {
	top: number
	right: number
	bottom: number
	left: number
}

/**
 * A context object for generating axis components
 */
export interface AxisContext {
	axis: Axis
	thickness: number
	scale: Scale<any, any>
	range: [number, number]
	rangeProperty: string
	rangeEndProperty: string
	crossProperty: string
	crossEndProperty: string
	horizontal: boolean
	topOrLeft: boolean
	frame: SceneFrame

	// Domain Data
	domainCross: number
	domainMinRange: number
	domainMaxRange: number

	// Tick Data
	ticks: PositionedTickValue[]
	tickCrossStart: number
	tickCrossEnd: number

	// Labels
	labelFormatter: (input: any) => string
}

export interface PositionedTickValue extends TickValue {
	position: number
}

/**
 * Interface for axis components such as ticks and labels
 */
export interface AxisComponent {
	/**
	 * Constructs the AxisContext object, which is used to render all axis components. Any
	 * pieces of context which this component generates or knows about should be populated here.
	 *
	 * @param context The intermediate context-object, which may be partially defined at this point
	 */
	createContext(context: Partial<AxisContext>): Partial<AxisContext>

	/**
	 * Determines whether the axis configuration results in this component emitting any information.
	 *
	 * @param context The axis context object
	 */
	isScenegraphElementGenerated(context: AxisContext): boolean

	/**
	 * Creates the scenegraph items for this axis component
	 *
	 * @param context The axis context object
	 */
	createScenegraphElement(context: AxisContext): SGMark<any>
}
