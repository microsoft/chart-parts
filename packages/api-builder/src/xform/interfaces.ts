/*!
 * Copyright (c) Microsoft. All rights reserved.
 * Licensed under the MIT license. See LICENSE file in the project.
 */

import { SGMark, SGItem, Channels } from '@chart-parts/interfaces'

/**
 * Interface for the result of scene-generation
 * @category Transformation
 */

export interface GeneratedScenegraph {
	/**
	 * The root mark of the resultant scene
	 */
	root: SGMark<SGItem>

	/**
	 * The event channels of this scene, a key of channel name to event handlers
	 */
	channelHandlers: Channels
}
