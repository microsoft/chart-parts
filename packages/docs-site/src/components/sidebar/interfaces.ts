/*!
 * Copyright (c) Microsoft. All rights reserved.
 * Licensed under the MIT license. See LICENSE file in the project.
 */

export interface SidebarItem {
	path: string
	title: string
	area: string
	order: number
}

export interface TreeNode {
	item: SidebarItem
	pathKey: string
	order?: number
	children: { [key: string]: TreeNode }
}
