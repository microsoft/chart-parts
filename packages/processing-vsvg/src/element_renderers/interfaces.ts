/*!
 * Copyright (c) Microsoft. All rights reserved.
 * Licensed under the MIT license. See LICENSE file in the project.
 */

import {
	MarkConverter,
	VSvgNode,
	VSvgTransformType,
} from '@chart-parts/interfaces'

export interface VSvgMarkOutput {
	nodes: VSvgNode[]
	defs?: VSvgNode[]
}

export interface VSvgRenderContext {
	nextId(): string
}

export type VSvgMarkConverter = MarkConverter<VSvgMarkOutput, VSvgRenderContext>

export function translate(x: number, y: number) {
	return { type: VSvgTransformType.translate, value: [x, y] }
}

export function rotate(angle: number) {
	return { type: VSvgTransformType.rotate, value: angle }
}
