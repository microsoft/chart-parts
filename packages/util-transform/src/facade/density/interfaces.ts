/*!
 * Copyright (c) Microsoft. All rights reserved.
 * Licensed under the MIT license. See LICENSE file in the project.
 */

import { DatasetManager } from '../dataset'
import { DatasetTransform } from '../interfaces'

export interface DistributionBuilder {
	build(ds: DatasetManager): any
}

/**
 * The density transform generates a new data stream of uniformly-spaced
 * samples drawn from a one-dimensional probability density function (pdf) or
 * cumulative distribution function (cdf). This transform is useful for
 * representing probability distributions and generating continuous
 * distributions from discrete samples using kernel density estimation.
 */
export interface DensityBuilder extends DatasetTransform {
	/**
	 * Required. An object describing the distribution type and parameters.
	 * See the distribution reference for more.
	 *
	 * @param dist
	 */
	distribution(dist: DistributionBuilder): DensityBuilder

	/**
	 * A [min, max] domain from which to sample the distribution. This argument
	 * is required in most cases, but can be omitted in the case of
	 * distributions (namely, kde) that can deduce their own extent.
	 *
	 * @param start
	 * @param end
	 */
	extent(start: number, end: number): DensityBuilder

	/**
	 * The type of distribution to generate.
	 * @param value
	 */
	method(value: DensityMethod): DensityBuilder

	/**
	 * The number of uniformly spaced steps to take along the extent domain
	 * (default 100). A total of steps + 1 uniformly-spaced samples are drawn
	 * from the distribution.
	 *
	 * @param value
	 */
	steps(value: number): DensityBuilder

	/**
	 * The output fields for the sample value and associated probability.
	 * The default is ["value", "density"].
	 *
	 * @param valueString
	 * @param probabilityField
	 */
	as(valueString: string, probabilityField: string): DensityBuilder
}

export enum DensityMethod {
	/**
	 * Probability Density Funciton
	 */
	pdf = 'pdf',

	/**
	 * Cumulative Distribution Function
	 */
	cdf = 'cdf',
}
