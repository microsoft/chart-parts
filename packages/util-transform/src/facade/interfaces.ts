/*!
 * Copyright (c) Microsoft. All rights reserved.
 * Licensed under the MIT license. See LICENSE file in the project.
 */

import { DatasetManager } from './dataset'

export interface DatasetTransform {
	/**
	 * Build the transformation
	 * @param df The vega dataflow instance
	 * @param from The vega-dataflow node that is the parent of this node
	 * @param ds The chart-parts dataset
	 */
	build(df: any, from: any, ds: DatasetManager): any
}
