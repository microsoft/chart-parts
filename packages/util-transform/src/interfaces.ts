/*!
 * Copyright (c) Microsoft. All rights reserved.
 * Licensed under the MIT license. See LICENSE file in the project.
 */

export type FieldAccessor = string // | ((row: any) => number)

export enum TextCaseTransform {
	lower = 'lower',
	upper = 'upper',
	mixed = 'mixed',
}

export enum CompareOrder {
	ascending = 'asc',
	descending = 'desc',
}

export interface Compare {
	field: FieldAccessor
	order?: CompareOrder
}

export enum Offset {
	zero = 'zero',
	center = 'center',
	normalize = 'normalize',
}

export type CompareFunction<T> = (a: T, b: T) => number
export type EqualityFunction<T> = (a: T, b: T) => boolean
export type GetKeyFunction<T> = (item: T) => any
export type Predicate<T> = (item: T) => boolean
export type Transformer<T, K> = (input: T) => K

export type MaybeNumber = number | undefined
