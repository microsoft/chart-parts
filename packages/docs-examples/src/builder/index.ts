/*!
 * Copyright (c) Microsoft. All rights reserved.
 * Licensed under the MIT license. See LICENSE file in the project.
 */
export * from './BarChartBuilder'
export * from './BarChartUtcBuilder'
export * from './GroupedBarChartBuilder'
