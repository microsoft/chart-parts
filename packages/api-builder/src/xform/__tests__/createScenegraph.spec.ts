/*!
 * Copyright (c) Microsoft. All rights reserved.
 * Licensed under the MIT license. See LICENSE file in the project.
 */

import { createScenegraph } from '../index'
import { scene, area } from '@chart-parts/builder'
import { ChartOptions } from '@chart-parts/interfaces'

describe('The createScenegraph() utility', () => {
	it('can transform a scene specification into a scenegraph', () => {
		const options: ChartOptions = { width: 250, height: 250 }
		const spec = scene(
			snb =>
				snb.mark(
					area()
						.table('data')
						.encode({
							x: ({ d }) => d.x,
							y: ({ d }) => d.y,
							y2: () => 0,
						}),
				),
			options,
		).build()

		const data = [
			{ x: 1, y: 50 },
			{ x: 2, y: 40 },
			{ x: 3, y: 30 },
			{ x: 4, y: 20 },
			{ x: 5, y: 10 },
		]
		const sg = createScenegraph(spec, { data }, options)

		expect(sg).toMatchSnapshot()
	})
})
