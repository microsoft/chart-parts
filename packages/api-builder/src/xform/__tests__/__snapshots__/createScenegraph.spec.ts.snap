// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The createScenegraph() utility can transform a scene specification into a scenegraph 1`] = `
Object {
  "channelHandlers": Object {},
  "root": Mark {
    "items": Array [
      Object {
        "channels": Object {},
        "cornerRadius": 0,
        "fill": "none",
        "height": 250,
        "items": Array [
          Mark {
            "items": Array [
              AreaItem {
                "channels": Object {},
                "fill": "none",
                "interpolate": "linear",
                "itemtype": "area",
                "metadata": Object {
                  "id": ".root.area.0",
                  "index": 0,
                },
                "name": undefined,
                "nodetype": "item",
                "orient": "vertical",
                "role": undefined,
                "stroke": "none",
                "strokeCap": "butt",
                "strokeJoin": "miter",
                "x": 1,
                "y": 50,
                "y2": 0,
                "zIndex": 0,
              },
              AreaItem {
                "channels": Object {},
                "fill": "none",
                "interpolate": "linear",
                "itemtype": "area",
                "metadata": Object {
                  "id": ".root.area.1",
                  "index": 1,
                },
                "name": undefined,
                "nodetype": "item",
                "orient": "vertical",
                "role": undefined,
                "stroke": "none",
                "strokeCap": "butt",
                "strokeJoin": "miter",
                "x": 2,
                "y": 40,
                "y2": 0,
                "zIndex": 0,
              },
              AreaItem {
                "channels": Object {},
                "fill": "none",
                "interpolate": "linear",
                "itemtype": "area",
                "metadata": Object {
                  "id": ".root.area.2",
                  "index": 2,
                },
                "name": undefined,
                "nodetype": "item",
                "orient": "vertical",
                "role": undefined,
                "stroke": "none",
                "strokeCap": "butt",
                "strokeJoin": "miter",
                "x": 3,
                "y": 30,
                "y2": 0,
                "zIndex": 0,
              },
              AreaItem {
                "channels": Object {},
                "fill": "none",
                "interpolate": "linear",
                "itemtype": "area",
                "metadata": Object {
                  "id": ".root.area.3",
                  "index": 3,
                },
                "name": undefined,
                "nodetype": "item",
                "orient": "vertical",
                "role": undefined,
                "stroke": "none",
                "strokeCap": "butt",
                "strokeJoin": "miter",
                "x": 4,
                "y": 20,
                "y2": 0,
                "zIndex": 0,
              },
              AreaItem {
                "channels": Object {},
                "fill": "none",
                "interpolate": "linear",
                "itemtype": "area",
                "metadata": Object {
                  "id": ".root.area.4",
                  "index": 4,
                },
                "name": undefined,
                "nodetype": "item",
                "orient": "vertical",
                "role": undefined,
                "stroke": "none",
                "strokeCap": "butt",
                "strokeJoin": "miter",
                "x": 5,
                "y": 10,
                "y2": 0,
                "zIndex": 0,
              },
            ],
            "marktype": "area",
            "metadata": Object {},
            "nodetype": "mark",
          },
        ],
        "itemtype": "group",
        "metadata": Object {
          "id": ".root.0",
          "index": 0,
        },
        "name": "root",
        "nodetype": "item",
        "role": "frame",
        "stroke": "none",
        "strokeCap": "butt",
        "strokeJoin": "miter",
        "width": 250,
        "x": 0,
        "x2": 250,
        "y": 0,
        "y2": 250,
        "zIndex": 0,
      },
    ],
    "marktype": "group",
    "metadata": Object {},
    "nodetype": "mark",
  },
}
`;
